### PR TITLE
feat(subsurface-view): Mixed layer creation.

### PR DESCRIPTION
**Pull request by LeonidPolukhin**
1) Additional SubsurfaceViewer prop added: **layerDefinitions**. It is an array containing layer definition Records as well as externally created layers. If provided, it is used instead of **layers** prop which is ignored. 
Behavior of **layers** is preserved for backward compatibility.

2) **@@typedArraySupport : true** in layer definition Record allows to use typed arrays as inputs bypassing JSON converter.

### CODE REVIEW

File: typescript/packages/subsurface-viewer/src/components/Map.tsx [hunk 1]
```diff
@@ -987,22 +989,43 @@ export function jsonToObject(
...
-        const layer = createLayer(layerData, configuration);
-        if (layer) {
-            layersList.push(layer);
+    // remove empty record/layer objects
+    const filtered_layers = layers.filter((value) => !isEmpty(value)) as (
+        | Record<string, unknown>
+        | Layer
+    )[];
+    const layersList = new Array(filtered_layers.length);
+    const jsonLayerIndices: number[] = [];
```
Comment (t0oF-azpn) [inline typescript/packages/subsurface-viewer/src/components/Map.tsx:1010]: I never like side by side arrays, whenever other solutions are possible (error prone, less readable).

can you use:
```
const jsonIndexedLayers: {layer: Record<string, unknown>, index: number}[];
...
jsonIndexedLayers.push({layer: layer, index: index});
```
But I let you choose to resolve this...

**Review by t0oF-azpn — CHANGES_REQUESTED** [review]

File: typescript/packages/subsurface-viewer/src/SubsurfaceViewer.tsx [hunk 2]
```diff
@@ -65,6 +72,15 @@ export interface SubsurfaceViewerProps {
     resources?: Record<string, unknown>;
     layers?: Record<string, unknown>[] | LayersList;
```
[the remark this answers is not in the document]
Comment (LeonidPolukhin) [reply typescript/packages/subsurface-viewer/src/SubsurfaceViewer.tsx:73]: The component is used from Python environment and I'm not sure that union type won't break anything there.

File: typescript/packages/subsurface-viewer/src/storybook/layers/Grid3DLayer.stories.tsx [hunk 3]
```diff
@@ -218,49 +220,54 @@ export const PolyhedralCells: StoryObj<typeof SubsurfaceViewer> = {
     parameters: parameters,
 };
 
-export const PolyhedralCellsTypedArrayInput: StoryObj<typeof SubsurfaceViewer> =
```
[the remark this answers is not in the document]
Comment (LeonidPolukhin) [reply typescript/packages/subsurface-viewer/src/storybook/layers/Grid3DLayer.stories.tsx:221]: Extracted to miscExamples

**Review by t0oF-azpn — APPROVED** [review]

**Comment by hkfb** [comment]
I might be wrong, but I think it should be possible to fix the current `layers` prop so that it would accept and work with both `Layer` and `Record<..>` types.

Moreover I don't get the use case for this. Seems like using `Layer` classes directly is better for type safety and to allow 3rd party layers.

**Comment by t0oF-azpn** [comment]
> I might be wrong, but I think it should be possible to fix the current `layers` prop so that it would accept and work with both `Layer` and `Record<..>` types.
> 
> Moreover I don't get the use case for this. Seems like using `Layer` classes directly is better for type safety and to allow 3rd party layers.

We agree on this, it's cool.
Leonid was concerned about possibly breaking the use with python: in this case, my understanding (thanks to Leonid) is that the plain javascript implementation gets used (so no more types), and the props are provided as a simple json string encoding the `Record<...>`. Which by the way explains the design with this json round-trip.

Leonid was waiting for your feedback before updating the PR to use only the current `layer`prop ;)

**Review by hkfb — APPROVED** [review]

**Pull request merged**